File: services/backend/nautionette_backend/reasoning.py
"""Provider-advertised reasoning controls, not guesses based on model names.

Copilot exposes capabilities.supports.reasoning_effort; OpenRouter exposes
reasoning.supported_efforts. A reasoning boolean alone does not define levels.
None is a provider default, never an instruction to disable reasoning.
"""

from typing import Any

EFFORTS = ("none", "minimal", "low", "medium", "high", "xhigh", "max")


def reasoning_metadata(item: dict[str, Any], integration_type: str) -> dict[str, Any]:
    capabilities = item.get("capabilities")
    supports = capabilities.get("supports", {}) if isinstance(capabilities, dict) else {}
    supports = supports if isinstance(supports, dict) else {}
    reasoning = item.get("reasoning")
    reasoning = reasoning if isinstance(reasoning, dict) else {}
    advertised = supports.get("reasoning_effort", reasoning.get("supported_efforts"))
    efforts = [effort for effort in EFFORTS if isinstance(advertised, list) and effort in advertised]
    supported = bool(efforts) if isinstance(advertised, list) else None
    if supported is None:
        if type(supports.get("reasoning_effort")) is bool:
            supported = supports["reasoning_effort"]
        elif reasoning:
            supported = True
    return {
        "supports_reasoning": supported,
        "reasoning_efforts": efforts,
        "reasoning_format": "openrouter" if integration_type == "openrouter" else "openai",
    }


def validate_effort(value: Any, model: dict[str, Any]) -> str | None:
    if value is None:
        return None
    if not isinstance(value, str) or value not in EFFORTS:
        raise ValueError("reasoning_effort must be null (provider default) or a supported effort level")
    if value not in model.get("reasoning_efforts", []):
        raise ValueError("This reasoning effort is not advertised for the selected model/API route")
    return value

File: images/agent-sets/default/extensions/nautionette/reasoning.ts
// The backend pins provider-advertised capabilities and the user's selection
// into each durable job. Do not infer capabilities from model names here.
const LEVELS = ["off", "minimal", "low", "medium", "high", "xhigh", "max"] as const;
type Level = typeof LEVELS[number];
const EFFORTS = ["none", ...LEVELS.slice(1)];

type ReasoningConfig = {
  effort: string | null;
  format: "openrouter" | "openai";
  reasoning: boolean;
  thinkingLevelMap: Partial<Record<Level, string | null>>;
  level: Level;
};

export function reasoningConfig(env: NodeJS.ProcessEnv = process.env): ReasoningConfig {
  const metadata = JSON.parse(env.NAUTIONETTE_MODEL_REASONING || "{}");
  const efforts = Array.isArray(metadata.efforts)
    ? EFFORTS.filter((effort) => metadata.efforts.includes(effort)) : [];
  const effort = env.NAUTIONETTE_REASONING_EFFORT || null;
  if (effort !== null && !efforts.includes(effort)) {
    throw new Error("Selected reasoning effort is not supported by the pinned model capabilities");
  }
  const format = metadata.format === "openrouter" ? "openrouter" : "openai";
  const thinkingLevelMap = Object.fromEntries(LEVELS.map((level) => [
    level, level === "off" ? "none" : efforts.includes(level) ? level : null,
  ]));
  return {
    effort,
    format,
    reasoning: metadata.supported === true || efforts.length > 0,
    thinkingLevelMap,
    level: (effort === null || effort === "none" ? "off" : effort) as Level,
  };
}

export function reasoningPayload(payload: Record<string, any>, api: string, config: ReasoningConfig) {
  const next = { ...payload };
  // Pi's off/default behavior can emit effort=none or enabled=false. Neither is
  // a provider default, and always-thinking models can reject them. Likewise,
  // Pi must not clamp max/xhigh or silently substitute a different effort.
  delete next.reasoning_effort;
  if (config.effort === null) {
    delete next.reasoning;
  } else if (api === "openai-responses" || config.format === "openrouter") {
    next.reasoning = { ...(config.effort === "none" ? {} : next.reasoning), effort: config.effort };
    delete next.reasoning.enabled;
  } else {
    delete next.reasoning;
    next.reasoning_effort = config.effort;
  }
  return next;
}
